src/NServiceBus.Core/Routing/Legacy/DetectDistributorDistributorBehavior.cs

```diff
@@ -0,0 +1,18 @@
+﻿namespace NServiceBus
+{
+    using System;
+    using System.Threading.Tasks;
+    using NServiceBus.Pipeline;
+
+    class DetectDistributorDistributorBehavior : Behavior<IIncomingLogicalMessageContext>
+    {
+        public override Task Invoke(IIncomingLogicalMessageContext context, Func<Task> next)
+        {
+            if (context.Headers.ContainsKey(LegacyDistributorHeaders.WorkerSessionId))
```

Comment:
could we save us a behavior here by checking the headers on the incoming message directly in the ApplyReplyToAddressBehavior?